crew happiness and memories

claudiusmueller opened the pull request:
The intent for this PR is to provide a framework for giving crew members more "life". It is intentionally **WIP** and provided here to help focus discussion on specific implementation details and whether or not we should pursue this.

My goal is to specify the smallest functional unit that provides at least some benefit to gameplay and can be built upon.

### Current roadmap (up for discussion)

- [x] Implement thoughts that impact relationship between crew member and player
- [x] Implement simple memory stack with decay + protection against thought spamming
- [X] Debug menu
- [X] Finalize detailed crew screen
- [X] Adjust detailed crew screen for player (no memories, happiness, etc.)
- [x] Adjust variables for good gameplay (memory decay time, memory stack size, impact on happiness per thought, etc.)

### Simple Summary

Player actions will directly impact the relationship individual crew members have with the player (aka. captain of the ship). If this "happiness" gets too low, crew members desert the player.

### Technical Details

Certain player actions trigger a thought in a crew member. Each thought has an immediate impact on the playerRelationship variable. 

Currently implemented thoughts:

- starting employment
- visiting the home station of the crew member
- trading illegal goods
- player being an offender/criminal/outlaw/fugitive
- visiting high population systems lately
- visiting low population systems lately
- visiting unexplored systems lately

The impact on the playerRelationship depends on the specific character of the crew member. For example, criminal activities will upset a crew member that is a law abiding citizen but may make a crew member with "loose moral standards" happy.

Once a thought is applied,  it receives a time stamp and is placed into the crew member's memory stack. Currently, the memory stack only holds four memories (up for debate). The memories decay over time (currently after a month) and are also safeguarded against spamming of the same memory (no more than twice the same thought in the memory at a time).

If the playerRelationship variable drops below threshold, crew members have a chance to desert the player at the next starport:

![crew_deserting_message](https://user-images.githubusercontent.com/51969417/231060353-0ba85e5d-c6cf-45f2-ac39-fbea503867a0.png)

They will not hire back on and, in fact, should make other candidates hesitant about joining. Still to be implemented: average crew happiness impact on willingness (price?) of new candidates joining.

I believe this will make the player responsible for the overall mood aboard his/her ship and also force them to select crew wisely.

To help players choose crew, the player can now gather some information about their interests in an "interview":

![hire_crew_interview_select](https://user-images.githubusercontent.com/51969417/231060494-231149ad-eb58-4524-8f6d-cbff0bb742ee.png)

They may not care about the law and/or be the adventurous type that wants to explore the unknown:

![hire_crew_interview_example_2](https://user-images.githubusercontent.com/51969417/231060582-8bb526a4-2406-451b-b65e-c1247073799a.png)

Or they may be a law abiding citizen and/or prefer visiting cultural centers:

![hire_crew_interview_example_3](https://user-images.githubusercontent.com/51969417/231060611-cb02d433-28e0-4ca9-9efd-a476a855a567.png)

Or maybe they don't care either way:

![hire_crew_interview_example_1](https://user-images.githubusercontent.com/51969417/231060650-dc226a80-d6ac-4e6d-9599-8aa03eede312.png)

Once crew is hired, the new crew screen would show details about their stats, their "happiness", and also their memory stack.

![crew_details_view_example_1](https://user-images.githubusercontent.com/51969417/231060702-53851bf7-0d97-4790-b669-457ca5dffe72.png)

### Choosing the right cutoffs for high/low system population impact on happiness

I used the following data to come up with the current cutoffs:

![system_pop_happiness_thresholds](https://user-images.githubusercontent.com/51969417/231060955-6c673426-563f-47de-9709-309978c1e989.png)

I also found this interesting:

![systems_per_cube](https://user-images.githubusercontent.com/51969417/231060993-8a2687e2-2673-435f-b240-657e88a52871.png)

Commit:
crew life initial commit

impaktor commented:
Some inspiration, from page 44 of GURPS Space 3ed

# CHARACTER TYPES

The literature of science fiction is so diverse that dozens
of character types are available for a Space campaign. We
discuss some of these below. Detailed character templates for
many character types can be found on pp. 87-105 of GURPS
Traveller.


## Starship Crew
Crew positions include Captain, First Officer, Pilot (or
Helmsman), Astrogator, Engineer, Surgeon, Technician,
Cargomaster, Steward, Gunner (or Weapons Officer), and
Communications, Science, and Security Officers. On smaller
ships, each crewman holds several positions; larger ships
have several crew for every position but Captain and First
Officer.

Skills of special value to starship crew include Armoury
(Spaceship Weaponry or Spaceship Armor), Astrogation,
Computer Operation, Electronics and Electronics Operation
(Communications, Force Shields, Medical, or Sensors),
Engineer (any space specialty), Free Fall, Freight Handling,
Gunner (any spaceship weapon), Mechanic (any space spe-
cialty), Piloting (any spaceship or auxiliary craft), and Vacc
Suit.

### Merchant
An interstellar trader who cruises the spaceways, buying
and selling. Merchant, Fast-Talk, and Diplomacy skills are a
must. Unless he’s a full-time trade specialist, he’ll also need
crew skills – especially on a small trading vessel.

### Navy
Crew on a regular military vessel. Marines are ground
troops transported via starship; see Trooper (p. 46). Savoir-
Faire (Military) skill is required, and officers should have
Leadership and Tactics. Military Rank is useful, unless you
want to be a private all your life.

### Patrolman
A member of the Patrol (see p. 21) or equivalent inter-
planetary agency, combining the skills of a police investiga-
tor with those of a soldier to keep the spacelanes safe. Sense
of Duty is a must.

### Pirate/Smuggler
The most celebrated interstellar criminals. Smugglers
run contraband from world to world – anything from guns to
drugs to slaves – slipping past the Patrol in fast, well-pro-
tected ships. They can use any of the Thief/Spy skills. Pirates
are the scourges of the spacelanes, attacking freighters, lin-
ers, lightly guarded colonies, low-tech worlds, prospectors,
and other prey. Some may be slavers. Combat/Weapon skills
and Odious Personal Habits are appropriate.

### Scout
Scouts (see p. 22) find and explore new worlds. They
might specialize in one or more sciences, or be “general spe-
cialists.” Useful skills include Cartography, Electronics
Operation (Sensors), Planetology, Survival, and
Xenobiology.

Commit:
remove old module file

Commit:
refactor and add debugging module

Commit:
remove old module file

Commit:
revert all changes

sturnclaw commented:
@claudiusmueller what is the size of the "cube" in the graphs you've provided? An `n` of 60 systems per sector seems quite high for sector generation, and I'm curious as to how you generated the 1d coordinate axis - was it `max(x, y, z)`?

claudiusmueller committed:
Merge branch 'pioneerspacesim:master' into master

Commit:
Merge branch 'master' into crewlife

Commit:
debug module for crewlife

claudiusmueller commented:
Added a debug module for crewlife to help develop and maintain.

<img width="1363" height="612" alt="debug_crewlife" src="https://github.com/user-attachments/assets/cbae47c0-dcfc-42a8-a2a3-fb2796e8dd18" />

Commit:
added thoughts for getting/not getting paid

Commit:
add event test onEnterSystem and fix PiGui IDs

claudiusmueller commented:
We may need some new icons at the indicated places:
1. for the debug menu tab
2. for good/bad memories

Or do you think the current market symbols used for good/bad memories are fine?

The good/bad memories icons would be used not just the debug screen but similarly on the crew info screen.

<img width="794" height="1123" alt="debug_crewlife_marked" src="https://github.com/user-attachments/assets/8aa4fb36-8045-4376-942c-a7b698b02cc4" />

Commit:
refactor and streamline crewlife, crewcontracts, crewdebug

Commit:
updated crew info view

claudiusmueller commented:
I need help with widget spacing for the crew info view. This screenshot shows the problem. The lines in question are found in info view 05-crew.lua, lines 348 - 364.

It looks like a the buttons are moved down to fill a pre-defined vertical space for PlayerInfoDetails. But the Vector2 for this ui child has y = 0, which should only fill as needed?

I'm sure it's something simple I'm overlooking here.

<img width="794" height="1123" alt="Screenshot_20260328_082427_space_debug_2_marked" src="https://github.com/user-attachments/assets/91c3d8b0-808b-4363-943a-1d355206d3bd" />

Commit:
fix widget naming

Commit:
only desert at busy stations

claudiusmueller commented:
Currently, the only real impact of this module (other than role play) is that crew members can desert the ship if morale is too low. To limit impact on game play I have limited this event to busy stations (pop > 0.4), which should have new crew available for hiring. This also makes sense for role play: why would you desert the ship if there is no reasonable expectation of a new job?

There are many more options here to integrate into game play, but I think it's better to include as a minimal module for now and get player feedback.

sturnclaw committed:
Improve crew UI display

- Two-column layout for crew stats using ui.child
- Move top-level layout of the crew inspect window into the main draw() handler.

sturnclaw commented:
> It looks like a the buttons are moved down to fill a pre-defined vertical space for PlayerInfoDetails. But the Vector2 for this ui child has y = 0, which should only fill as needed?

Y = 0 means "fill all available space". It cannot reserve only the space needed for its contents, because the contents of the child have not been submitted yet. Generally speaking, you should not use multiple nested child windows - that is the source of your problem.

There are several other major issues with the way the drawCrewInfo function is computed.
- `region_column_width` is the same as `info_column_width`, because you're calling `ui.getContentRegion()` *before* you've entered the info panel child window (which is the action that restricts the content region to the left half of the screen).
- You have a great number of unnecessary `ui.child` calls, the contents of which should just be submitted regularly to the top-level window. If you're dividing the window vertically, you can continue to submit to the left column before submitting to the right column.
- Not necessarily an execution problem, but you have mixed spaces and tabs in the file, and you're using 2-space indentation which is *not* the convention Pioneer uses. You'll need to normalize those before the PR can be accepted.

I'm pushing a commit with a quick set of fixes to illustrate the correct layout and architecture, hopefully it helps you to build from there.

Additionally, you have a typo in `CrewLife.lua#190` which is causing `Timer:CallAt(...)` to fail since the function isn't being assigned to the right value. This causes a runtime crash when hiring crew...

Commit:
fix and simplify crew ui

Commit:
fix widget spacing

Commit:
remove superflous check

Commit:
fix crew hire crash

Commit:
full crew hire process via debug menu

Commit:
fix indentation

bszlrd commented:
@claudiusmueller How about these icons?
<img width="85" height="41" alt="image" src="https://github.com/user-attachments/assets/1d4fb4e5-623a-42ff-bc62-45931a16fc6d" />
Both are from Tabler. Storm cloud for bad memories, sun for happy ones.
Will these be shown to the player too, or just in debug?

I'd suggest that if there will be color indication, then use the scheme the ship market has: 
<img width="618" height="363" alt="image" src="https://github.com/user-attachments/assets/dff5940d-d170-4d95-a66e-94cd78f21acb" />
Much more colorblind friendly than green-red

claudiusmueller commented:
@bszlrd Thanks - I'll give these a try and also change the color scheme. They are visible to the player, not just the debug.

bszlrd commented:
Here's the svg
You'd still need to set them up in the theme though
<img width="640" height="840" alt="icons-2026feb" src="https://github.com/user-attachments/assets/f2affde1-fcd9-43b9-a091-34ae3d495121" />

claudiusmueller commented:
Screenshots with updated color theme and icons. I loaded the icons temporarily as images, so inline alignment is slightly off (but that fixes itself if I use an actual icon).

## Debug window:
<img width="1524" height="778" alt="crewlife_update_1" src="https://github.com/user-attachments/assets/936e751e-08a5-49bd-a700-6080eda11b64" />

## Crew view:
<img width="2557" height="1278" alt="crewlife_update_2" src="https://github.com/user-attachments/assets/92aaee61-cfbf-4a3a-80e7-a76004fb5bf4" />


## Still need to select an icon for the debug tab header here:
<img width="794" height="1123" alt="crewlife_update_3_marked" src="https://github.com/user-attachments/assets/f54a2e4e-cfb6-4f5e-a002-68ecad3c058e" />


### My suggestion would be this (from tabler): 
<img width="24" height="24" alt="users-colorized" src="https://github.com/user-attachments/assets/b8f2a248-1543-45f5-9d3c-64658a4b47fb" />
I can't test with this as image though because the debugView expects an actual icon.

bszlrd commented:
<img width="640" height="840" alt="icons-2026feb" src="https://github.com/user-attachments/assets/f1ed6da5-2299-4bb5-9f00-bf0cde56ac81" />

Commit:
update icons, theme + cleanup

Commit:
fix variable name

claudiusmueller commented:
Added new icons, color theme, and did some script cleanup:

# debug menu
<img width="1522" height="735" alt="crewlife_update_4" src="https://github.com/user-attachments/assets/b7a0bb1f-941f-4c1d-9aca-a6e3c6c72a22" />

# crew roster view
<img width="2556" height="1353" alt="crewlife_update_5" src="https://github.com/user-attachments/assets/0adeb1ac-5091-4938-8c25-b968dfb2df15" />

Commit:
pick and show home station for each crew member

claudiusmueller commented:
Added proper selection of home system per crew member. 

Highest probability = station the potential crew member is hired from, followed by stations within the same system (by distance) and then surrounding systems (by distance) up to 30 ly away.

<img width="2546" height="1434" alt="crewlife_update_6" src="https://github.com/user-attachments/assets/da7351f5-1e95-4380-bed5-90db0e4dc9d4" />

bszlrd commented:
Hehe, there could be a slight chance of them originating from somewhere further away. And there could be a chance that they'd quit when you land in that system, partly based on their job satisfaction. So in essence instead of paying for passage, they'd work for it, but in a shady way.

claudiusmueller commented:
Yeah - that's what I'm trying to simulate with the Poisson distribution based selection. Normal = hired at home, with decreasing probability of further away. 

Cool idea though about working for passage. I'll write that down. Maybe even secretly working for passage (will desert once you land at their home)?

Commit:
better crew home station algorithm

claudiusmueller commented:
Changed the algorithm that picks crew home station. It now selects:

1. ~60% local system, 40% outside of local system (99% within 10 ly)
2. within local system: 40% current station, then decreasing probability with increased distance from current station

Commit:
set and track last home visit

claudiusmueller commented:
Added tracking of last home visit for each crew member.

<img width="2045" height="1150" alt="crewlife_update_7" src="https://github.com/user-attachments/assets/3e6ffca9-1623-4ea9-937c-bbc40f18f250" />


<img width="1504" height="839" alt="crewlife_update_8" src="https://github.com/user-attachments/assets/556905d7-a802-4b28-9f7e-bccbac11ae03" />

Commit:
var call typo

Commit:
docstrings and cleanup

Commit:
Streamline findHome using MissionUtils

claudiusmueller commented:
Streamlined findHome algorithm using MissionUtils functions. Here is a snapshot of the base selection distribution for 1000 crew members. This can be easily tweaked with a Poisson lambda.
<img width="543" height="534" alt="crewlife_update_9" src="https://github.com/user-attachments/assets/62a19492-4682-45e7-b026-5f31650bcb28" />

Commit:
fix crash due to old save incompatibility

claudiusmueller commented:
I'll do more play testing - but this is ready for review.

impaktor commented:
There's a merge conflict that needs to be resolved.

Suggestion: In the "date since last visit to home system", could you also add in parenthesis a (X days ago), or similar? 
Also makes me wonder what the player should feel / realize when seeing it. Is there a "oh, this guy hasn't been home in a while!" thought? Or is it "this guy can do 2 years in deep space standing on his head"? If they're home sick, you could also add color coding to the ("X days") since last visit home?

claudiusmueller commented:
@impaktor - time delta has been added (see screenshots). Currently, home sickness has not been implemented. I want to introduce this gently... The crew member does get a morality boost though if you visit their home and it has been at least 30 days ago (which you can also see in their thoughts/memories).

I have to admit - I have difficulty understanding the impact of the individual icons.svg merge conflicts.

<img width="2047" height="1150" alt="crewlife_update_10" src="https://github.com/user-attachments/assets/d44ec4c6-24b4-4f31-9bd7-25fed98b7b0d" />

<img width="1503" height="820" alt="crewlife_update_11" src="https://github.com/user-attachments/assets/1d8d9455-77db-4384-a41d-c45b4febc656" />

Commit:
add time delta to home visit

bszlrd commented:
Okay, I think the merge conflict is because of #6339, the fuel pump clarification PR  (Heh, I like that issue number)
I've also noticed that at least my browser mangled the svg, [so here it is on dropbox](https://www.dropbox.com/scl/fi/531ykybrzopy5865lhm5k/icons.svg?rlkey=b101i5a75hh9s1id5wdj508nd&st=ydy57474&dl=0), just in case.

Commit:
fixed negative home visit dates and stronger bias for local home

claudiusmueller commented:
Fixed a silly error placing home visit dates into the future. Also adjusted the home station selection to bias the home station more.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into crewlife

Commit:
fixed crash when creating crew in space

claudiusmueller commented:
Found a bug during play testing. For some odd reason I thought onUpdateBB is only run when a player is docking (or after some time while docked). This fix accounts for updated BB's while the player is flying.

Commit:
removed leftover commented code

Commit:
fix crash when paying crew
